consensus/istanbul/uptime/autofix.go

```diff
@@ -0,0 +1,174 @@
+	}
+
+	lastHeaderNumber := lastHeader.Number.Uint64()
+
+	// If it's a rewind, rebuild
+	if number < lastHeaderNumber {
+		af.builder.Clear()
+		return af.cleanBuild(header)
+	}
+
+	// Normal sequential case
+	if number == lastHeaderNumber+1 {
+		if header.ParentHash == lastHeader.Hash() {
+			// Valid chain, add the header
+			return af.builder.ProcessHeader(header)
+		}
+		// Fork! Rebuild
+		af.builder.Clear()
+		return af.cleanBuild(header)
+	}
+
+	if number > lastHeaderNumber+1 {
+		// Normal advance (1 or more headers missing). Try to advance the builder
+		return af.advance(lastHeader, header)
+	}
+
+	// number == lastHeaderNumber
+	// check if it's an idempotent call or if
+	// there' a fork
+	if lastHeader.Hash() == header.Hash() {
+		// Nothing to do.
+		return nil
+	}
+
+	// fork, rebuild
+	af.builder.Clear()
+	return af.cleanBuild(header)
+}
+
+// cleanBuild does a clean build up to the header given, assuming the decorated builder is empty.
+func (af *autoFixBuilder) cleanBuild(upTo *types.Header) error {
+	epochSize := af.builder.GetEpochSize()
+	numberWithinEpoch := istanbul.GetNumberWithinEpoch(upTo.Number.Uint64(), epochSize)
+	// If this is the first header in the epoch, no need to call the provider
+	if numberWithinEpoch == 1 {
+		return af.builder.ProcessHeader(upTo)
+	}
+	// Else, load all the previous headers from the db.
+	headers, err := af.provider.GetEpochHeadersUpToLimit(epochSize, upTo, numberWithinEpoch)
+	if err != nil {
+		return err
+	}
+	return af.addAll(headers)
+}
+
+// advance advances the computation from the header given to the upTo header given.
+// This method requests to the provider the missing headers in between and adds them.
+// The only caveat is if the 'from' header is not in the same chain as the 'upTo' header.
+// In that case, a rebuild is made.
+func (af *autoFixBuilder) advance(from, upTo *types.Header) error {
+	upNumber := upTo.Number.Uint64()
+	fromNumber := from.Number.Uint64()
+	limit := upNumber - fromNumber + 1
+	headers, err := af.provider.GetEpochHeadersUpToLimit(af.builder.GetEpochSize(), upTo, limit)
+	if err != nil {
+		return err
+	}
+	// Verify that this is not a fork
+	if headers[0].Hash() != from.Hash() {
+		// Fork!
+		// Rebuild
+		af.builder.Clear()
+		return af.cleanBuild(upTo)
+	}
+	return af.addAll(headers[1:])
+}
+
+// addAll adds all the headers to the decorated builder.
+func (af *autoFixBuilder) addAll(headers []*types.Header) error {
+	for _, h := range headers {
+		err := af.builder.ProcessHeader(h)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (af *autoFixBuilder) copy() *autoFixBuilder {
+	return &autoFixBuilder{
+		builder:  af.builder.Copy(),
+		provider: af.provider,
+	}
+}
+
+func (af *autoFixBuilder) ComputeUptime(epochLastHeader *types.Header) ([]*big.Int, error) {
+	// We use a deep copy for ComputeUptime, to maintain the postcondition
+	// of ComputeUptime (the builder will not be modified).
+	b := af.copy()
```

Comment:
potential improvement is to check if `ProcessHEader` is required. For example if epochLastHeader is already LastHeader or if it outside the monitoring window, then we don't need to call ProcessHeader, and we avoid copying

Reply:
I think that the idea is not breaking the actual accumulator just because of the compute. The deep copy is not a hard one and it will only be triggered when this compute is called. I don't see a scenario where we are calling this for every block, and if that's the case, I prefer to leave the algorithm less tied of epoch conditions. Agree that it's an optimization, but we are adding more restrictions instead of making it simpler